modules/layers/src/icon-layer/icon-layer.js

```diff
@@ -118,19 +131,33 @@ export default class IconLayer extends Layer {
     /* eslint-enable max-len */
   }
 
+  _updateIconMapping(iconMapping) {
+    const attributeManager = this.getAttributeManager();
+    attributeManager.invalidate('instanceOffsets');
+    attributeManager.invalidate('instanceIconFrames');
+    attributeManager.invalidate('instanceColorModes');
+    this.setState({iconMapping});
+  }
+
   updateState({oldProps, props, changeFlags}) {
     super.updateState({props, oldProps, changeFlags});
 
     const {iconAtlas, iconMapping} = props;
+    const {iconManager} = this.state;
 
-    if (oldProps.iconMapping !== iconMapping) {
-      const attributeManager = this.getAttributeManager();
-      attributeManager.invalidate('instanceOffsets');
-      attributeManager.invalidate('instanceIconFrames');
-      attributeManager.invalidate('instanceColorModes');
+    if (iconManager && changeFlags.dataChanged) {
+      iconManager.setData(props.data);
+      if (this.state.iconMapping !== iconManager.mapping) {
+        this._updateIconMapping(iconManager.mapping);
+        this.setState({iconsTexture: iconManager.texture});
+      }
+    }
+
+    if (iconMapping && oldProps.iconMapping !== iconMapping) {
+      this._updateIconMapping(iconMapping);
```

Comment:
This will overwrite the result of line 151, if the user mistakenly supplies this prop.